### PR TITLE
Spark: Better SparkBatchScan statistics estimation

### PR DESCRIPTION
**Pull request by wypoon**
When estimating statistics, we should take the read schema into account.
As explained in the [PR](https://github.com/apache/spark/pull/33825) for [SPARK-36568](https://issues.apache.org/jira/browse/SPARK-36568):
`V2ScanRelationPushDown` can column prune `DataSourceV2ScanRelation`s and change read schema of `Scan` operations. 
In `SparkBatchScan.estimateStatistics()`, before this change, for unpartitioned tables or partitioned tables with no filter expressions, we sum the file sizes of the files to be scanned and use that as the estimate for the size in bytes. With this change, we adjust that by the ratio of the size of the columns to be read to the size of all the columns; we also adjust that by the compression factor in case it is set (the default is 1.0). With this change, some joins that are broadcast joins on V1 tables but are SortMergeJoins on Iceberg tables can be broadcast joins as well.
Basically this change does for Iceberg what the above-mentioned Spark PR does for the built-in V2 `FileScan`s.

### CODE REVIEW

**Comment by rdblue** [comment]
I'm surprised to see that we report the total data file size back to Spark. At Netflix, we switched over to reporting row size estimate * number of rows a long time ago to fix broadcast join memory problems. Spark should ignore the size on disk and use the row size * num rows, but if Spark doesn't do that then I think Iceberg should report that as the size instead.

**Comment by rdblue** [comment]
Thanks @wypoon! I'm okay with this, but I'd prefer to return a better estimate based on number of rows and not compressed size at all. Interested to hear what @aokolnychyi thinks.

File: spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java [hunk 1]
```diff
@@ -280,23 +280,23 @@ private static PartitionSpec identitySpec(Schema schema, List<String> partitionN
...
     }
 
-    long approximateSize = 0;
-    for (StructField sparkField : tableSchema.fields()) {
-      approximateSize += sparkField.dataType().defaultSize();
+    long result;
+    try {
+      result = LongMath.checkedMultiply(tableSchema.defaultSize(), totalRecords);
+    } catch (ArithmeticException e) {
+      result = Long.MAX_VALUE;
```
Comment (wypoon) [inline spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java:298]: `StructType` has a `defaultSize` method; there is no need to re-implement it. The main utility of the `estimateSize` static method is checking for overflow; we should just use Guava's `LongMath.checkedMultiply`.

File: spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchScan.java [hunk 2]
```diff
@@ -207,22 +207,20 @@ public Statistics estimateStatistics() {
       LOG.debug("using table metadata to estimate table statistics");
       long totalRecords = PropertyUtil.propertyAsLong(table.currentSnapshot().summary(),
           SnapshotSummary.TOTAL_RECORDS_PROP, Long.MAX_VALUE);
-      Schema projectedSchema = expectedSchema != null ? expectedSchema : table.schema();
```
Comment (wypoon) [inline spark3/src/main/java/org/apache/iceberg/spark/source/SparkBatchScan.java:210]: I don't think `expectedSchema` is ever null. Thus, I think we can just call `readSchema()` to get the `StructType`.

**Comment by wypoon** [comment]
@rdblue thanks for reviewing this. I changed the statistics estimation to use (row size * number of rows) for all cases. Of course, the row size estimation can be very far from accurate. Is this version preferable? @aokolnychyi or @RussellSpitzer do you prefer this change?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @wypoon! Looks great.

**Comment by aokolnychyi** [comment]
Sorry for the delay. I think this change makes sense too. Thanks, @wypoon!

Shall we update Spark 2 as well?

**Comment by aokolnychyi** [comment]
I created #3108 so that we don't forget.

**Comment by wypoon** [comment]
@aokolnychyi thanks for seconding the change.
Yes, I think it makes sense to do the same for Spark 2 as well. I'll put up a PR.

**Comment by aokolnychyi** [comment]
Much appreciated, @wypoon!

**Comment by zinking** [comment]
> Thanks @wypoon! I'm okay with this, but I'd prefer to return a better estimate based on number of rows and not compressed size at all. Interested to hear what @aokolnychyi thinks.

well, rows * schema size over estimates table sizes under some circumstances, for example TPCDS-sf1000 Q7 demographies dim table, causing broadcast join degrade to sort merge join.

I guess it still works in 3.2 because of AQE. 

totalSize * readcols size / total cols size  is what hive adopted. but certainly this is underestimating in some circumstances.
